### DISCUSSION TITLE
不能分布式编译

### DISCUSSION THREAD

**Discussion by Domain** [xmake-io/xmake#2485] [Q&A] [answered]
最新的2.6.8，服务器上执行xmake service --distcc --start，提示成功，配置文件也生成了。客户端配置client.conf，确保IP和token没错，执行xmake service --connect --distcc，提示连接成功。再执行xmake，只提示一句话，<remote_build_client>: run xmake in 192.168.1.100:9691，然后没反应了。看起来像是执行了远程编译，但是也没输出。客户端没在编译。另外，一开始会上传所有源码，这是不是没必要？工程比较大，上传花了很久，一些没必要的文件也上传了，例如生成的VS工程。fastbuild是在本地进行预处理，然后分发处理好的文件给各个客户端编译，不需要上传全部。

**Comment by Domain** [comment]
用过-vD选项，客户端还是没有任何输出，服务器log显示编译失败。因为有些源代码是需要自动生成，这些文件本地已经生成了，但是并没有上传到服务器。另外日志文件里面包含了很多颜色控制符，看起来很不直观

**Comment by waruqi** [comment, marked as answer]
那是因为连到 远程编译的 服务上去了。。没有连到 分布式编译服务。。。

`xmake service --distcc --start` 先不要开 --start ，去掉后，加 `-vD` ，在前台开下符号，看下服务 logs 。。然后贴过来

```console
xmake service --distcc -vD
```

然后，客户端删了 `rm -rf .xmake` ，然后重连下 `xmake service --connect --distcc -vD` 试试，也加上 -vD 。。给下完整 logs

另外，确认下两端的端口是否一致。。分布式服务端口是 9693 ，远程编译服务是 9691 。。端口不要配置错误

**Comment by Domain** [comment]
我重新试了一下，用xmake service --distcc -vD启动服务器，再xmake service --connect --distcc -vD，服务器和客户端都提示成功，再执行xmake的时候，提示server unreachable。然后我把client.conf里面的remote_build和remote_cache的配置删除，只保留distcc_build，再执行xmake，提示config(remote_build.connect): not found! 感觉还是远程编译，没有分布式编译

**Reply by waruqi** [reply]
凡事先贴 logs

**Comment by Domain** [comment]
```
C:\Users\build>xmake service --distcc -vD
<distcc_build_server>: listening 0.0.0.0:9693 ..
<distcc_build_server>: <socket: tcp4/808.0>: session connected
<distcc_build_server>: <socket: tcp4/808.0>: <session 5958f8c1>: on handle message(2)
{
  session_id = "5958f8c1",
  code = 2,
  token = "36e8ee0a7ec5e12f382183e7b885eae7"
}

<distcc_build_server>: <socket: tcp4/808.0>: <session 5958f8c1>: send ok
<distcc_build_server>: <socket: tcp4/808.0>: session end
<distcc_build_server>: <socket: tcp4/884.0>: session connected
<distcc_build_server>: <socket: tcp4/884.0>: <session 5958f8c1>: on handle message(1)
{
  code = 1,
  session_id = "5958f8c1",
  token = "36e8ee0a7ec5e12f382183e7b885eae7",
  xmakever = "2.6.8"
}

<distcc_build_server>: <socket: tcp4/884.0>: <session 5958f8c1>: send ok
<distcc_build_server>: <socket: tcp4/884.0>: session end
```

**Comment by Domain** [comment]
C:\build>xmake -vD
error: @programdir\core\main.lua:280: @programdir\modules\private\service\remote_build\client.lua:267: <remote_build_client>: server unreachable!
stack traceback:
    [C]: in function 'error'
    [@programdir\core\base\os.lua:872]: in function 'raiselevel'
    [@programdir\core\sandbox\modules\utils.lua:181]: in function 'assert'
    [@programdir\modules\private\service\remote_build\client.lua:267]: in function 'runcmd'
    [@programdir\modules\private\service\remote_build\action.lua:32]:
    [C]: in function 'xpcall'
    [@programdir\core\base\utils.lua:280]:
    [@programdir\core\base\task.lua:501]: in function 'run'
    [@programdir\core\main.lua:278]: in function 'cotask'
    [@programdir\core\base\scheduler.lua:388]:

stack traceback:
        [C]: in function 'error'
        @programdir\core\base\os.lua:872: in function 'os.raiselevel'
        (...tail calls...)
        @programdir\core\main.lua:280: in upvalue 'cotask'
        @programdir\core\base\scheduler.lua:388: in function <@programdir\core\base\scheduler.lua:385>

**Reply by waruqi** [reply]
> C:\build>xmake -vD error: @programdir\core\main.lua:280: @programdir\modules\private\service\remote_build\client.lua:267: <remote_build_client>: server unreachable! stack traceback: [C]: in function 'error' [@programdir\core\base\os.lua:872]: in function 'raiselevel' [@programdir\core\sandbox\modules\utils.lua:181]: in function 'assert' [@programdir\modules\private\service\remote_build\client.lua:267]: in function 'runcmd' [@programdir\modules\private\service\remote_build\action.lua:32]: [C]: in function 'xpcall' [@programdir\core\base\utils.lua:280]: [@programdir\core\base\task.lua:501]: in function 'run' [@programdir\core\main.lua:278]: in function 'cotask' [@programdir\core\base\scheduler.lua:388]:
> 
> stack traceback: [C]: in function 'error' @programdir\core\base\os.lua:872: in function 'os.raiselevel' (...tail calls...) @programdir\core\main.lua:280: in upvalue 'cotask' @programdir\core\base\scheduler.lua:388: in function <@programdir\core\base\scheduler.lua:385>

> 然后，客户端删了 rm -rf .xmake ，然后重连下 xmake service --connect --distcc -vD 试试，也加上 -vD 。。给下完整 logs

没看到相关操作和 logs

**Reply by Domain** [reply]
删除.xmake之后可以了，虽然还是有一句错误提示：
error config(remote_build.connect): not found
不过后面看到有distcc在编译了

**Comment by waruqi** [comment]
> error config(remote_build.connect): not found

那是因为你刚删了 相关配置，先不要去删

**Reply by Domain** [reply]
我已经加回来了
